docs/cleanup-design.md

```diff
@@ -0,0 +1,40 @@
+|    of BDx       +-------->+                  +-------->+                  +-------->+               | Successfully|            |
+|                 |         |                  |         |                  |         |               |             |            |
++-----------------+         +------------------+         +--------+---------+         +-------+-------+             +------+-----+
+                                                                  |                           |                            |	
+                                                                  | Yes                       |Not Found                   |	
+                                                                  v                           v                            |	
+                                                          +-------+---------+          +------+-------+                    |	
+                             +-----------------+          |                 |          |              |                    |	
+                             |                 |    No    |  Check if BD is |          |   Check if   |                    |	
+                             | Cancel the job  +<---------+  in active state|          | BD is active |                    |	
+                             |                 |          |                 |          |              |                    |	
+                             |                 |          |                 |          |              |                    |	
+                             +-------+---------+          +-----------------+          +-------+------+                    |	
+                                     |                             |                           |Yes                    	   |	
+                                     |                             |                    +------+-----+                     |	
+                                     |                             |                    |            |                     |	
+                                     |                             | Yes                | start job  |                     |	
+                                     v                             |                    |            |                     |	
+                              +------+---------+                   |                    |            |                     |	
+                              |                |                   |                    +-------+----+                     |	
+                              |    return      |<------------------+----------------------------|--------------------------+ 	
+                              +----------------+                                    		  			 
+                                                                                                            	                 
+```
```

Comment:
Does a cancel job delete the job or mark it as completed? What happens in the following scenario:
- BD is released. 
- Cleanup Job is scheduled
- During the Cleanup, BD goes inactive - due to node failure or other errors
- ndm-operator sees that BD is inactive and cancels the job. 
--
- After a while the node is back online
- ndm-operator checks the status of the BD is released. and looks for a job. Finds a job that is cancelled/completed? Will the BD remain in released state?

Reply:
@kmova  A cancel will delete the job, without checking its status (can be completed/running/pending etc).

Workflow for the mentioned sequence of events:
1. BD is released
2. Cleanup Job is scheduled
3. During cleanup, BD goes inactive - due to some reasons
4. NDM-Operator sees that BD is inactive and cancels the job (Delete the job without checking status). The BD status will not be updated.  (will be `Released` itself)
5. In the next reconcile loop of the BD, operator finds that the BD is in released state and no jobs are present for the BD. But it will not start the job since the BD is inactive.

- Step 5 will happen in case of all subsequent reconcile loops till BD becomes `Active`

6. After sometime, node is back online and BD is `Active`
7. In the next reconcile loop, Operator schedules the cleanup job again, since the BD is in `Released` state and is `Active`
8. Once the cleanup job is executed successfully (i.e job marked completed by kubernetes), the job is deleted and BD is marked as `Unclaimed`